# COMS W4705 Spring 24
## Homework 4 - Semantic Role Labelling with BERT

The goal of this assignment is to train and evaluate a PropBank-style semantic role labeling (SRL) system. Following (Collobert et al. 2011) and others, we will treat this problem as a sequence-labeling task. For each input token, the system will predict a B-I-O tag, as illustrated in the following example: 

|The|judge|scheduled|to|preside|over|his|trial|was|removed|from|the|case|today|.|             
|---|-----|---------|--|-------|----|---|-----|---|-------|----|---|----|-----|-|             
|B-ARG1|I-ARG1|B-V|B-ARG2|I-ARG2|I-ARG2|I-ARG2|I-ARG2|O|O|O|O|O|O|O|
|||schedule.01|||||||||||||

Note that the same sentence may have multiple annotations for different predicates

|The|judge|scheduled|to|preside|over|his|trial|was|removed|from|the|case|today|.|             
|---|-----|---------|--|-------|----|---|-----|---|-------|----|---|----|-----|-|             
|B-ARG1|I-ARG1|I-ARG1|I-ARG1|I-ARG1|I-ARG1|I-ARG1|I-ARG1|O|B-V|B-ARG2|I-ARG2|I-ARG2|B-ARGM-TMP|O|
||||||||||remove.01||||||

and not all predicates need to be verbs

|The|judge|scheduled|to|preside|over|his|trial|was|removed|from|the|case|today|.|             
|---|-----|---------|--|-------|----|---|-----|---|-------|----|---|----|-----|-|    
|O|O|O|O|O|O|B-ARG1|B-V|O|O|O|O|O|O|O|
||||||||try.02||||||||

The SRL system will be implemented in [PyTorch](https://pytorch.org/). We will use BERT (in the implementation provided by the [Huggingface transformers](https://huggingface.co/docs/transformers/index) library) to compute contextualized token representations and a custom classification head to predict semantic roles. We will fine-tune the pretrained BERT model on the SRL task. 


### Overview of the Approach

The model we will train is pretty straightforward. Essentially, we will just encode the sentence with BERT, then take the contextualized embedding for each token and feed it into a classifier to predict the corresponding tag. 

Because we are only working on argument identification and labeling (not predicate identification), it is essentially that we tell the model where the predicate is. This can be accomplished in various ways. The approach we will choose here repurposes Bert's *segment embeddings*. 

Recall that BERT is trained on two input sentences, seperated by [SEP], and on a next-sentence-prediction objective (in addition to the masked LM objective). To help BERT comprehend which sentence a given token belongs to, the original BERT uses a segment embedding, using A for the first sentene, and B for the second sentence 2. 
Because we are labeling only a single sentence at a time, we can use the segment embeddings to indicate the predicate position instead: The predicate is labeled as segment B (1) and all other tokens will be labeled as segment A (0). 

<img src="https://github.com/daniel-bauer/4705-f23-hw5/blob/main/bert_srl_model.png?raw=true" width=400px>

## Setup: GCP, Jupyter, PyTorch, GPU

To make sure that PyTorch is available and can use the GPU,run the following cell which should return True. If it doesn't, make sure the GPU drivers and CUDA are installed correctly. 

GPU support is required for this assignment -- you will not be able to fine-tune BERT on a CPU. 

In [ ]:
import torch
torch.cuda.is_available() 

## Dataset: Ontonotes 5.0 English SRL annotations

We will work with the English part of the [Ontonotes 5.0](https://catalog.ldc.upenn.edu/LDC2013T19) data. This is an extension of PropBank, using the same type of annotation. Ontonotes contains annotations other than predicate/argument structures, but we will use the PropBank style SRL annotations only. *Important*: This data set is provided to you for use in COMS 4705 only! Columbia is a subscriber to LDC and is allowed to use the data for educational purposes. However, you may not use the dataset in projects unrelated to Columbia teaching or research.

If you haven't done so already, you can download the data here: 


In [52]:
! wget https://storage.googleapis.com/4705-bert-srl-data/ontonotes_srl.zip    

--2024-12-11 23:21:55--  https://storage.googleapis.com/4705-bert-srl-data/ontonotes_srl.zip
Resolving storage.googleapis.com (storage.googleapis.com)... 142.250.68.59, 142.250.68.123, 142.250.72.187, ...
Connecting to storage.googleapis.com (storage.googleapis.com)|142.250.68.59|:443... connected.
HTTP request sent, awaiting response... 

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)


200 OK
Length: 12369688 (12M) [application/zip]
Saving to: ‘ontonotes_srl.zip.3’

ontonotes_srl.zip.3 100%[===================>]  11.80M  21.3MB/s    in 0.6s    

2024-12-11 23:21:56 (21.3 MB/s) - ‘ontonotes_srl.zip.3’ saved [12369688/12369688]



In [53]:
! unzip ontonotes_srl.zip    

Archive:  ontonotes_srl.zip
replace propbank_dev.tsv? [y]es, [n]o, [A]ll, [N]one, [r]ename: 

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)


^C


The data has been pre-processed in the following format. There are three files: 

`propbank_dev.tsv`	`propbank_test.tsv`	`propbank_train.tsv`

Each of these files is in a tab-separated value format. A single predicate/argument structure annotation consists of four rows. For example 

```
ontonotes/bc/cnn/00/cnn_0000.152.1
The     judge   scheduled       to      preside over    his     trial   was     removed from    the     case    today   /.
                schedule.01
B-ARG1  I-ARG1  B-V     B-ARG2  I-ARG2  I-ARG2  I-ARG2  I-ARG2  O       O       O       O       O       O       O
```

* The first row is a unique identifier (1st annotation of the 152nd sentence in the file ontonotes/bc/cnn/00/cnn_0000).
* The second row contains the tokens of the sentence (tab-separated). 
* The third row contains the probank frame name for the predicate (empty field for all other tokens). 
* The fourth row contains the B-I-O tag for each token. 

The file `rolelist.txt` contains a list of propbank BIO labels in the dataset (i.e. possible output tokens). This list has been filtered to contain only roles that appeared more than 1000 times in the training data. 
We will load this list and create mappings from numeric ids to BIO tags and back.

In [149]:
role_to_id = {}
with open("role_list.txt",'r') as f:
    role_list = [x.strip() for x in f.readlines()]
    role_to_id = dict((role, index) for (index, role) in enumerate(role_list))
    role_to_id['[PAD]'] = -100
    
    id_to_role = dict((index, role) for (role, index) in role_to_id.items())
    

Note that we are also mapping the '[PAD]' token to the value -100. This allows the loss function to ignore these tokens during training.

## Part 1 - Data Preparation

Before you can build the SRL model, you first need to preprocess the data. 


### 1.1 - Tokenization 

One challenge is that the pre-trained BERT model uses subword ("WordPiece") tokenization, but the Ontonotes data does not. Fortunately Huggingface transformers provides a tokenizer. 

In [150]:
from transformers import BertTokenizerFast 
tokenizer = BertTokenizerFast.from_pretrained('bert-base-uncased', do_lower_case=True)
tokenizer.tokenize("This is an unbelievably boring test sentence.")

['this',
 'is',
 'an',
 'un',
 '##bel',
 '##ie',
 '##va',
 '##bly',
 'boring',
 'test',
 'sentence',
 '.']

**TODO**: 
We need to be able to maintain the correct labels (B-I-O tags) for each of the subwords. 
Complete the following function that takes a list of tokens and a list of B-I-O labels of the same length as parameters, and returns a new token / label pair, as illustrated in the following example. 


```
>>> tokenize_with_labels("the fancyful penguin devoured yummy fish .".split(), "B-ARG0 I-ARG0 I-ARG0 B-V B-ARG1 I-ARG1 O".split(), tokenizer)
(['the',
  'fancy',
  '##ful',
  'penguin',
  'dev',
  '##oured',
  'yu',
  '##mmy',
  'fish',
  '.'],
 ['B-ARG0',
  'I-ARG0',
  'I-ARG0',
  'I-ARG0',
  'B-V',
  'I-V',
  'B-ARG1',
  'I-ARG1',
  'I-ARG1',
  'O'])

```

To approach this problem, iterate through each word/label pair in the sentence. Call the tokenizer on the word. This may result in one or more tokens. Create the correct number of labels to match the number of tokens. Take care to not generate multiple B- tokens. 


This approach is a bit slower than tokenizing the entire sentence, but is necessary to produce proper input tokenization for the pre-trained BERT model, and the matching target labels. 

In [151]:
def tokenize_with_labels(sentence, text_labels, tokenizer):
    """
    Word piece tokenization makes it difficult to match word labels
    back up with individual word pieces. 
    """

    tokenized_sentence = []
    labels = []
    
    for word, label in zip(sentence, text_labels):
        w_toks = tokenizer.tokenize(word)
        tokenized_sentence.append(w_toks[0])
        labels.append(label)
        for tok in w_toks[1:]:
            tokenized_sentence.append(f'##{tok}')
            if label != 'O':
                labels.append(f'I-{label[2:]}')
            else:
                labels.append('O')

    return tokenized_sentence, labels

In [152]:
tokenize_with_labels("the fancyful penguin devoured yummy fish .".split(), "B-ARG0 I-ARG0 I-ARG0 B-V B-ARG1 I-ARG1 O".split(), tokenizer)

(['the',
  'fancy',
  '####ful',
  'penguin',
  'dev',
  '####oured',
  'yu',
  '####mmy',
  'fish',
  '.'],
 ['B-ARG0',
  'I-ARG0',
  'I-ARG0',
  'I-ARG0',
  'B-V',
  'I-V',
  'B-ARG1',
  'I-ARG1',
  'I-ARG1',
  'O'])

### 1.2 Loading the Dataset

Next, we are creating a PyTorch [Dataset](https://pytorch.org/docs/stable/data.html#torch.utils.data.Dataset) class. This class acts as a contained for the training, development, and testing data in memory. You should already be familiar with Datasets and Dataloaders from homework 3. 

1.2.1 **TODO**: Write the \_\_init\_\_(self, filename) method that reads in the data from a data file (specified by the filename).

For each annotation you start with  the tokens in the sentence, and the BIO tags. Then you need to create the following 

1. call the `tokenize_with_labels` function to tokenize the sentence.
2. Add the (token, label) pair to the self.items list. 

1.2.2 **TODO**: Write the \_\_len\_\_(self) method that returns the total number of items. 

1.2.3 **TODO**: Write the \_\_getitem\_\_(self, k) method that returns a single item in a format BERT will understand. 
* We need to process the sentence by adding "\[CLS\]" as the first token and "\[SEP\]" as the last token. The need to pad the token sequence to 128 tokens using the "\[PAD\]" symbol. This needs to happen both for the inputs (sentence token sequence) and outputs (BIO tag sequence).
* We need to create an *attention mask*, which is a sequence of 128 tokens indicating the actual input symbols (as a 1) and \[PAD\] symbols (as a 0).
* We need to create a *predicate indicator* mask, which is a sequence of 128 tokens with at most one 1, in the position of the "B-V" tag. All other entries should be 0. The model will use this information to understand where the predicate is located. 

* Finally, we need to convert the token and tag sequence into numeric indices. For the tokens, this can be done using the `tokenizer.convert_tokens_to_ids` method. For the tags, use the `role_to_id` dictionary. 
Each sequence must be a pytorch tensor of shape (1,128). You can convert a list of integer values like this `torch.tensor(token_ids, dtype=torch.long)`.

To keep everything organized, we will return a dictionary in the following format 

```
{'ids': token_tensor,
 'targets': tag_tensor,
 'mask': attention_mask_tensor,
 'pred': predicate_indicator_tensor}
```


(Hint: To debug these, read in the first annotation only / the first few annotations) 


In [189]:
from torch.utils.data import Dataset, DataLoader 
import torch

def load_SRL_data(filename):
    with open(filename,'r') as prop_list_f: 
        return [line.strip() for line in prop_list_f] 
   

class SrlData(Dataset):
    
    def __init__(self, filename):

        super(SrlData, self).__init__()
        
        self.max_len = 128 # the max number of tokens inputted to the transformer. 
        
        self.tokenizer = BertTokenizerFast.from_pretrained('bert-base-uncased', do_lower_case=True)        
        
        self.items = []
        
        data = load_SRL_data(filename)
        for s,props in zip(data[1::4],data[3::4]):
            toks_labels = tokenize_with_labels(s.split(), props.split(), self.tokenizer)
            self.items.append(toks_labels)
                                                        
    def __len__(self):
        return len(self.items)
    
    def __getitem__(self, k):
        toks,labels = self.items[k]
        s_len = len(toks)

        if s_len > self.max_len - 2:  # -2 for [CLS] and [SEP]
            toks = toks[:(self.max_len - 2)]
            labels = labels[:(self.max_len - 2)]
            s_len = self.max_len - 2
        
        toks = ['[CLS]'] + toks + ['[SEP]'] + (['[PAD]'] * (128 - (s_len + 2)))
        toks = self.tokenizer.convert_tokens_to_ids(toks)
        toks = torch.tensor(toks,dtype=torch.long).unsqueeze(0)
        
        labels = ['[CLS]'] + labels + ['[SEP]'] + (['[PAD]'] * (128 - (s_len + 2)))
        tags = []
        for l in labels:
            if l not in role_to_id:
                tags.append(role_to_id['O'])
            else:
                tags.append(role_to_id[l])
        tags = torch.tensor(tags,dtype=torch.long).unsqueeze(0)
        
        mask = ([1] * (s_len + 2)) + ([0] * (128 - (s_len + 2)))
        mask = torch.tensor(mask,dtype=torch.long).unsqueeze(0)
        
        pred_mask = [0] * 128
        if 'B-V' in labels:
            pred_mask[labels.index('B-V')] = 1
        pred_mask = torch.tensor(pred_mask,dtype=torch.long)
        
        return {'ids': toks,
                'mask':  mask,
                'targets': tags,
                'pred': pred_mask
               }
        

In [190]:
# Reading the training data takes a while for the entire data because we preprocess all data offline
data = SrlData("propbank_train.tsv")

## 2. Model Definition

In [191]:
from torch.nn import Module, Linear, CrossEntropyLoss
from transformers import BertModel

We will define the pyTorch model as a subclass of the [torch.nn.Module](https://pytorch.org/docs/stable/generated/torch.nn.Module.html) class. The code for the model is provided for you. It may help to take a look at the documentation to remind you of how Module works. Take a look at how the huggingface BERT model simply becomes another sub-module. 

In [192]:
class SrlModel(Module):
    
    def __init__(self):
        
        super(SrlModel, self).__init__()
        
        self.encoder = BertModel.from_pretrained("bert-base-uncased")
    
        # The following two lines would freeze the BERT parameters and allow us to train the classifier by itself.
        # We are fine-tuning the model, so you can leave this commented out!
        # for param in self.encoder.parameters():
        #    param.requires_grad = False
        
        # The linear classifier head, see model figure in the introduction. 
        self.classifier = Linear(768, len(role_to_id))
                
        
    def forward(self, input_ids, attn_mask, pred_indicator):
    
        # This defines the flow of data through the model 
    
        # Note the use of the "token type ids" which represents the segment encoding explained in the introduction. 
        # In our segment encoding, 1 indicates the predicate, and 0 indicates everything else. 
        bert_output =  self.encoder(input_ids=input_ids, attention_mask=attn_mask, token_type_ids=pred_indicator)

        enc_tokens = bert_output[0] # the result of encoding the input with BERT
        logits = self.classifier(enc_tokens) #feed into the classification layer to produce scores for each tag.
        
        # Note that we are only interested in the argmax for each token, so we do not have to normalize 
        # to a probability distribution using softmax. The CrossEntropyLoss loss function takes this into account.
        # It essentially computes the softmax first and then computes the negative log-likelihood for the target classes. 
        return logits        

In [193]:
model = SrlModel().to('cuda') # create new model and store weights in GPU memory

Now we are ready to try running the model with just a single input example to check if it is working correctly. Clearly it has not been trained, so the output is not what we expect. But we can see what the loss looks like for an initial sanity check. 

**TODO**: 
* Take a single data item from the dev set, as provided by your Dataset class defined above. Obtain the input token ids, attention mask, predicate indicator mask, and target labels. 
* Run the model on the ids, attention mask, and predicate mask like this: 

In [194]:
# pick an item from the dataset. Then run 
dev_data = SrlData("propbank_dev.tsv")

In [195]:
ix = 10
data_dict = dev_data[ix]

ids = data_dict['ids'].to('cuda')
mask = data_dict['mask'].to('cuda')
pred = data_dict['pred'].to('cuda')
targets = data_dict['targets'].to('cuda')
print(f'ids: {ids.shape}, mask: {mask.shape}, pred: {pred.shape}, targs: {targets.shape}')

outputs = model(ids, mask, pred)
outputs.shape

ids: torch.Size([1, 128]), mask: torch.Size([1, 128]), pred: torch.Size([128]), targs: torch.Size([1, 128])


torch.Size([1, 128, 53])

**TODO**: 
Compute the loss on this one item only.
The initial loss should be close to -ln(1/num_labels)

Without training we would assume that all labels for each token (including the target label) are equally likely, so the negative log probability for the targets should be approximately $$-\ln(\frac{1}{\text{num_labels}}).$$ This is what the loss function should return on a single example. This is a good sanity check to run for any multi-class prediction problem. 

In [196]:
import math
-math.log(1 / len(role_to_id), math.e)

3.970291913552122

In [197]:
loss_function = CrossEntropyLoss(ignore_index = -100, reduction='mean')

# Transpose outputs to get shape [1, 53, 128]
outputs = outputs.transpose(1, 2)

# Now targets is [1, 128] which is correct
loss = loss_function(outputs, targets)
print(f"Loss: {loss.item()}")

# complete this. Note that you still have to provide a (batch_size, input_pos) 
# tensor for each parameter, where batch_size =1

# outputs = model(ids, mask, pred) 
# loss = loss_function(...)
# loss.item()   #this should be approximately the score from the previous cell


Loss: 4.078204154968262


**TODO**: At this point you should also obtain the actual predictions by taking the argmax over each position.
The result should look something like this (values will differ).

```
tensor([[ 1,  4,  4,  4,  4,  4,  5, 29, 29, 29,  4, 28,  6, 32, 32, 32, 32, 32,
         32, 32, 30, 30, 32, 30, 32,  4, 32, 32, 30,  4, 49,  4, 49, 32, 30,  4,
         32,  4, 32, 32,  4,  2,  4,  4, 32,  4, 32, 32, 32, 32, 30, 32, 32, 30,
         32,  4,  4, 49,  4,  4,  4,  4,  4,  4,  4,  4,  4,  4,  6,  6, 32, 32,
         30, 32, 32, 32, 32, 32, 30, 30, 30, 32, 30, 49, 49, 32, 32, 30,  4,  4,
          4,  4, 29,  4,  4,  4,  4,  4,  4, 32,  4,  4,  4, 32,  4, 30,  4, 32,
         30,  4, 32,  4,  4,  4,  4,  4, 32,  4,  4,  4,  4,  4,  4,  4,  4,  4,
          4,  4]], device='cuda:0')
```

Then use the id_to_role dictionary to decode to actual tokens. 

```
['[CLS]', 'O', 'O', 'O', 'O', 'O', 'B-ARG0', 'I-ARG0', 'I-ARG0', 'I-ARG0', 'O', 'B-V', 'B-ARG1', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG1', 'I-ARG1', 'I-ARG2', 'I-ARG1', 'I-ARG2', 'O', 'I-ARG2', 'I-ARG2', 'I-ARG1', 'O', 'I-ARGM-TMP', 'O', 'I-ARGM-TMP', 'I-ARG2', 'I-ARG1', 'O', 'I-ARG2', 'O', 'I-ARG2', 'I-ARG2', 'O', '[SEP]', 'O', 'O', 'I-ARG2', 'O', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG1', 'I-ARG2', 'I-ARG2', 'I-ARG1', 'I-ARG2', 'O', 'O', 'I-ARGM-TMP', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'B-ARG1', 'B-ARG1', 'I-ARG2', 'I-ARG2', 'I-ARG1', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG1', 'I-ARG1', 'I-ARG1', 'I-ARG2', 'I-ARG1', 'I-ARGM-TMP', 'I-ARGM-TMP', 'I-ARG2', 'I-ARG2', 'I-ARG1', 'O', 'O', 'O', 'O', 'I-ARG0', 'O', 'O', 'O', 'O', 'O', 'O', 'I-ARG2', 'O', 'O', 'O', 'I-ARG2', 'O', 'I-ARG1', 'O', 'I-ARG2', 'I-ARG1', 'O', 'I-ARG2', 'O', 'O', 'O', 'O', 'O', 'I-ARG2', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O']
```

For now, just make sure you understand how to do this for a single example. Later, you will write a more formal function to do this once we have trained the model. 

In [261]:
# Get predictions by taking argmax along the class dimension
predictions = torch.argmax(outputs, dim=1)  # Shape will be [1, 128]

tok_preds = []
for ix, n in enumerate(predictions.to('cpu').tolist()[0]):
    if n not in id_to_role:
        tok_preds.append('O')
    else:
        tok_preds.append(id_to_role[n])

print(tok_preds)

['I-ARGM-ADV', 'I-ARGM-LVB', 'B-ARGM-DIS', 'B-ARGM-DIS', 'B-ARGM-DIS', 'B-ARGM-DIS', 'B-ARG4', 'B-ARGM-DIS', 'B-ARGM-DIS', 'O', 'B-ARG4', 'B-ARGM-DIS', 'B-ARGM-DIS', 'I-ARG1', 'B-ARGM-MNR', 'B-ARGM-MNR', 'I-ARG2', 'I-ARG2', 'I-ARG1', 'I-ARG2', 'I-ARG2', 'I-ARG1', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG1', 'O', 'I-ARG1', 'I-ARGM-GOL', 'I-ARGM-ADV', 'I-ARG2', 'I-ARGM-NEG', 'B-ARG4', 'I-ARG2', 'I-ARGM-NEG', 'I-ARG2', 'I-ARG2', 'I-ARGM-NEG', 'I-ARG2', 'I-ARG2', 'I-ARGM-NEG', 'I-ARG1', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG1', 'B-ARGM-DIS', 'O', 'I-ARG1', 'I-ARGM-GOL', 'I-ARGM-ADV', 'I-ARG2', 'B-ARG4', 'I-ARGM-ADV', 'I-ARG2', 'I-ARGM-NEG', 'B-ARG4', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'B-ARG4', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG2', 'I-ARG1', 'B-ARGM-DIS', 'I-ARG2', 'I-ARGM-

## 3. Training loop

pytorch provides a DataLoader class that can be wrapped around a Dataset to easily use the dataset for training. The DataLoader allows us to easily adjust the batch size and shuffle the data. 

In [199]:
from torch.utils.data import DataLoader
loader = DataLoader(data, batch_size = 32, shuffle = True)

The following cell contains the main training loop. The code should work as written and report the loss after each batch,
cumulative average loss after each 100 batches, and print out the final average loss after the epoch. 

**TODO**: Modify the training loop belowso that it also computes the accuracy for each batch and reports the 
average accuracy after the epoch. 
The accuracy is the number of correctly predicted token labels out of the number of total predictions. 
Make sure you exclude [PAD] tokens, i.e. tokens for which the target label is -100. It's okay to include [CLS] and [SEP] in the accuracy calculation. 

In [202]:
loss_function = CrossEntropyLoss(ignore_index = -100, reduction='mean')

LEARNING_RATE = 1e-05
optimizer = torch.optim.AdamW(params=model.parameters(), lr=LEARNING_RATE)

device = 'cuda'

def train():
    """
    Train the model for one epoch.
    """
    tr_loss = 0 
    nb_tr_examples, nb_tr_steps = 0, 0
    tr_preds, tr_labels = [], []
    total_correct = 0
    total_predictions = 0
    # put model in training mode
    model.train()
    
    for idx, batch in enumerate(loader):
        
        # Get the encoded data for this batch and push it to the GPU
        ids = batch['ids'].squeeze(1).to(device, dtype = torch.long)
        mask = batch['mask'].squeeze(1).to(device, dtype = torch.long)
        targets = batch['targets'].squeeze(1).to(device, dtype = torch.long)
        pred_mask = batch['pred'].squeeze(1).to(device, dtype = torch.long)

        # Run the forward pass of the model
        logits = model(input_ids=ids, attn_mask=mask, pred_indicator=pred_mask)        
        loss = loss_function(logits.transpose(2,1), targets) 
        tr_loss += loss.item()
        # print("Batch loss: ", loss.item()) # can comment out if too verbose.
        
        nb_tr_steps += 1
        nb_tr_examples += targets.size(0)
        
        if idx % 100==0:
            #torch.cuda.empty_cache() # can help if you run into memory issues
            curr_avg_loss = tr_loss/nb_tr_steps
            print(f"Current average loss: {curr_avg_loss}")
                   
        # Compute accuracy for this batch
        matching = torch.sum(torch.argmax(logits,dim=2) == targets)       
        predictions = torch.sum(torch.where(targets==-100,0,1))

        total_correct += matching.item()
        total_predictions += predictions.item()
                
        # Run the backward pass to update parameters
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

    epoch_loss = tr_loss / nb_tr_steps
    epoch_accuracy = total_correct / total_predictions if total_predictions > 0 else 0

    print(f"Training loss epoch: {epoch_loss}")
    print(f"Training accuracy epoch: {epoch_accuracy}")


Now let's train the model for one epoch. This will take a while (up to a few hours). 

In [203]:
train()

Current average loss: 1.199844241142273
Current average loss: 1.0573418429582426
Current average loss: 0.9984575824357977
Current average loss: 0.9563420975327096
Current average loss: 0.9265619123963049
Current average loss: 0.8936250694021731
Current average loss: 0.8631760304462096
Current average loss: 0.8368185120455378
Current average loss: 0.8148368098092883
Current average loss: 0.793070558768663
Current average loss: 0.7743656497139793
Current average loss: 0.7555619941636933
Current average loss: 0.7388871813147987
Current average loss: 0.7228814854152746
Current average loss: 0.7079119443638167
Current average loss: 0.6930224598645369
Current average loss: 0.6795355472097093
Current average loss: 0.6674891628197962
Current average loss: 0.6560261631627535
Current average loss: 0.6455117541667601
Current average loss: 0.6346194347043683
Current average loss: 0.6247922250099377
Current average loss: 0.6150932489864288
Current average loss: 0.6065080439148132
Current average lo

In my experiments, I found that two epochs are needed for good performance. 

In [204]:
train() 

Current average loss: 0.20046962797641754
Current average loss: 0.21293375456687247
Current average loss: 0.21131860550066725
Current average loss: 0.2134070033052831
Current average loss: 0.21675453045198745
Current average loss: 0.21910934199294643
Current average loss: 0.22131952986790615
Current average loss: 0.21978585602723583
Current average loss: 0.21823122545834933
Current average loss: 0.21760978063487849
Current average loss: 0.2173287448111352
Current average loss: 0.21729202023779232
Current average loss: 0.21656204972859724
Current average loss: 0.2162424916606239
Current average loss: 0.2164528869507272
Current average loss: 0.21538340291138652
Current average loss: 0.21469378297093286
Current average loss: 0.2141739760297211
Current average loss: 0.21340179423915884
Current average loss: 0.21376472136602473
Current average loss: 0.2133763051558768
Current average loss: 0.2132294177920378
Current average loss: 0.21317673801719086
Current average loss: 0.21289104323790467

I ended up with a training loss of about 0.19 and a training accuracy of 0.94. Specific values may differ. 

At this point, it's a good idea to save the model (or rather the parameter dictionary) so you can continue evaluating the model without having to retrain. 

In [206]:
torch.save(model.state_dict(), "srl_model_fulltrain_2epoch_finetune_1e-05.pt")

## 4. Decoding

In [ ]:
# Optional step: If you stopped working after part 3, first load the trained model 

model = SrlModel().to('cuda') 
model.load_state_dict(torch.load("srl_model_fulltrain_2epoch_finetune_1e-05.pt"))
model = model.to('cuda')

/tmp/ipykernel_10969/3484726954.py:4: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.
  model.load_state_dict(torch.load("srl_model_fulltrain_2epoch_finetune_1e-05.pt"))


**TODO (this is the fun part)**: Now that we have a trained model, let's try labeling an unseen example sentence. Complete the functions decode_output and label_sentence below. decode_output takes the logits returned by the model, extracts the argmax to obtain the label predictions for each token, and then translate the result into a list of string labels. 

label_sentence takes a list of input tokens and a predicate index, prepares the model input, call the model and then call decode_output to produce a final result. 

Note that you have already implemented all components necessary (preparing the input data from the token list and predicate index, decoding the model output). But now you are putting it together in one convenient function. 

In [207]:
tokens = "A U. N. team spent an hour inside the hospital , where it found evident signs of shelling and gunfire .".split()

In [238]:
def decode_output(logits): # it will be useful to have this in a separate function later on
    """
    Given the model output, return a list of string labels for each token. 
    """
    # Get predictions by taking argmax along the class dimension
    predictions = torch.argmax(logits, dim=1)  # Shape will be [1, 128]
    print(predictions)

    tok_preds = []
    for ix, n in enumerate(predictions.to('cpu').tolist()[0]):
        tok_preds.append(id_to_role[n])

    return tok_preds
    

In [243]:
def label_sentence(tokens, pred_idx):
    
    # complete this function to prepare token_ids, attention mask, predicate mask, then call the model. 
    # Decode the output to produce a list of labels. 
    
    model.eval()
    
    tokenizer = BertTokenizerFast.from_pretrained('bert-base-uncased', do_lower_case=True)
    bert_tokens = []
    token_map = []  # Maps original tokens to BERT token positions
    
    for i, word in enumerate(tokens):
        word_tokens = tokenizer.tokenize(word)
        bert_tokens.extend(word_tokens)
        token_map.extend([i] * len(word_tokens))
        
    # Truncate if needed
    if len(bert_tokens) > 126:  # Reserve space for [CLS] and [SEP]
        bert_tokens = bert_tokens[:126]
        token_map = token_map[:126]
    
    # Add special tokens
    bert_tokens = ['[CLS]'] + bert_tokens + ['[SEP]']
    token_map = [-1] + token_map + [-1]  # -1 for special tokens
    
    input_ids = tokenizer.convert_tokens_to_ids(bert_tokens)
    padding = [tokenizer.pad_token_id] * (128 - len(input_ids))
    input_ids.extend(padding)
    
    attention_mask = [1] * len(bert_tokens) + [0] * (128 - len(bert_tokens))
    
    # Create predicate mask (token_type_ids)
    pred_mask = [0] * 128
    bert_pred_positions = [i for i, x in enumerate(token_map) if x == pred_idx]
    if bert_pred_positions:
        pred_mask[bert_pred_positions[0]] = 1  # Mark first subword of predicate
    
    # Convert to tensors and move to GPU
    input_ids = torch.tensor([input_ids], dtype=torch.long).to(device)
    attention_mask = torch.tensor([attention_mask], dtype=torch.long).to(device)
    pred_mask = torch.tensor([pred_mask], dtype=torch.long).to(device)
    
    with torch.no_grad():
        outputs = model(input_ids=input_ids, attn_mask=attention_mask, pred_indicator=pred_mask)
    
    predictions = torch.argmax(outputs, dim=2)[0].cpu().numpy()
    final_labels = []
    
    for i, token in enumerate(tokens):
        # Find all BERT tokens corresponding to this original token
        bert_indices = [j for j, x in enumerate(token_map) if x == i]
        if not bert_indices:
            final_labels.append("O")
            continue
            
        # Use the prediction of the first subword token
        pred_id = predictions[bert_indices[0]]
        if pred_id in id_to_role:
            label = id_to_role[pred_id]
            
            # Handle continuation of B- labels
            if (i > 0 and label.startswith('B-') and 
                len(final_labels) > 0 and  # Make sure we have previous labels
                (final_labels[-1].startswith('B-') or final_labels[-1].startswith('I-')) and
                final_labels[-1][2:] == label[2:]):  # Check if it's the same role type
                    label = 'I-' + label[2:]
            final_labels.append(label)
        else:
            final_labels.append("O")
    
    return final_labels

In [242]:
# Now you should be able to run

label_test = label_sentence(tokens, 13) # Predicate is "found"
for i,j in zip(tokens, label_test):
    print(f'{i}: {j}')

A: O
U.: O
N.: O
team: O
spent: O
an: O
hour: O
inside: O
the: B-ARGM-LOC
hospital: I-ARGM-LOC
,: O
where: B-ARGM-LOC
it: B-ARG0
found: B-V
evident: B-ARG1
signs: I-ARG1
of: I-ARG1
shelling: I-ARG1
and: I-ARG1
gunfire: I-ARG1
.: O


The expected output is somethign like this: 
```   
 ('A', 'O'),
 ('U.', 'O'),
 ('N.', 'O'),
 ('team', 'O'),
 ('spent', 'O'),
 ('an', 'O'),
 ('hour', 'O'),
 ('inside', 'O'),
 ('the', 'B-ARGM-LOC'),
 ('hospital', 'I-ARGM-LOC'),
 (',', 'O'),
 ('where', 'B-ARGM-LOC'),
 ('it', 'B-ARG0'),
 ('found', 'B-V'),
 ('evident', 'B-ARG1'),
 ('signs', 'I-ARG1'),
 ('of', 'I-ARG1'),
 ('shelling', 'I-ARG1'),
 ('and', 'I-ARG1'),
 ('gunfire', 'I-ARG1'),
 ('.', 'O'),
```


### 5. Evaluation 1: Token-Based Accuracy
We want to evaluate the model on the dev or test set. 

In [217]:
dev_data = SrlData("propbank_dev.tsv") # Takes a while because we preprocess all data offline

In [218]:
from torch.utils.data import DataLoader
loader = DataLoader(dev_data, batch_size = 1, shuffle = False)

In [ ]:
# Optional: Load the model again if you stopped working prior to this step. 
# model = SrlModel()
# model.load_state_dict(torch.load("srl_model_fulltrain_2epoch_finetune_1e-05.pt"))
# model = mode.to('cuda')

**TODO**: Complete the evaluate_token_accuracy function below. The function should iterate through the items in the data loader (see training loop in part 3). Run the model on each sentence/predicate pair and extract the predictions.

For each sentence, count the correct predictions and the total predictions. Finally, compute the accuracy as #correct_predictions / #total_predictions

Careful: You need to filter out the padded positions ([PAD] target tokens), as well as [CLS] and [SEP]. It's okay to include [B-V] in the count though. 

In [247]:
def evaluate_token_accuracy(model, loader):
    
    model.eval()  # Set model to evaluation mode
    
    total_correct = 0
    total_predictions = 0
    
    with torch.no_grad():  # Disable gradient computation for efficiency
        for batch in loader:
            # Get batch data and move to GPU
            ids = batch['ids'].squeeze(1).to('cuda')
            mask = batch['mask'].squeeze(1).to('cuda')
            targets = batch['targets'].squeeze(1).to('cuda')
            pred_mask = batch['pred'].squeeze(1).to('cuda')
            
            # Get model predictions
            outputs = model(input_ids=ids, attn_mask=mask, pred_indicator=pred_mask)
            predictions = torch.argmax(outputs, dim=2)
            
            # Create mask for valid tokens (excluding [PAD], [CLS], [SEP])
            valid_tokens = targets != -100  # -100 is the padding index
            
            # Count correct predictions
            correct = (predictions == targets) & valid_tokens
            total_correct += correct.sum().item()
            total_predictions += valid_tokens.sum().item()
    
    accuracy = total_correct / total_predictions if total_predictions > 0 else 0
    print(f"Token Accuracy: {accuracy:.4f}")
    print(f"Total correct predictions: {total_correct}")
    print(f"Total valid tokens: {total_predictions}")
    
    return accuracy
dev_loader = DataLoader(dev_data, batch_size=32, shuffle=False)
accuracy = evaluate_token_accuracy(model, dev_loader)
print(accuracy)

Token Accuracy: 0.9358
Total correct predictions: 1299700
Total valid tokens: 1388939
0.9357502381314082


### 6. Span-Based evaluation 

While the accuracy score in part 5 is encouraging, an accuracy-based evaluation is problematic for two reasons. First, most of the target labels are actually O. Second, it only tells us that per-token prediction works, but does not directly evaluate the SRL performance. 

Instead, SRL systems are typically evaluated on micro-averaged precision, recall, and F1-score for predicting labeled spans. 

More specifically, for each sentence/predicate input, we run the model, decode the output, and extract a set of labeled spans (from the output and the target labels). These spans are (i,j,label) tuples.  

We then compute the true_positives, false_positives, and false_negatives based on these spans. 

In the end, we can compute 

* Precision:  true_positive / (true_positives + false_positives)  , that is the number of correct spans out of all predicted spans. 

* Recall: true_positives / (true_positives + false_negatives) , that is the number of correct spans out of all target spans. 

* F1-score:   (2 * precision * recall) / (precision + recall)


For example, consider 

| |[CLS]|The|judge|scheduled|to|preside|over|his|trial|was|removed|from|the|case|today|.|             
|--||---|-----|---------|--|-------|----|---|-----|---|-------|----|---|----|-----|-|             
||0|1|2|3|4|5|6|7|8|9|1O|11|12|13|14|15|
|target|[CLS]|B-ARG1|I-ARG1|B-V|B-ARG2|I-ARG2|I-ARG2|I-ARG2|I-ARG2|O|O|O|O|O|O|O|
|prediction|[CLS]|B-ARG1|I-ARG1|B-V|I-ARG2|I-ARG2|O|O|O|O|O|O|O|O|B-ARGM-TMP|O|

The target spans are (1,2,"ARG1"), and (4,8,"ARG2").

The predicted spans would be (1,2,"ARG1"), (14,14,"ARGM-TMP"). Note that in the prediction, there is no proper ARG2 span because we are missing the B-ARG2 token, so this span should not be created. 

So for this sentence we woudl get: true_positives: 1 false_positives: 1 false_negatives: 1

*TODO*: Complete the function evaluate_spans that performs the span-based evaluation on the given model and data loader. You can use the provided extract_spans function, which returns the spans as a dictionary. For example
{(1,2): "ARG1", (4,8):"ARG2"} 

In [ ]:
def extract_spans(labels):
    spans = {} # map (start,end) ids to label
    current_span_start = 0
    current_span_type = ""
    inside = False
    for i, label in enumerate(labels):
        if label.startswith("B"):            
            if inside: 
                if current_span_type != "V":
                    spans[(current_span_start,i)] = current_span_type            
            current_span_start = i
            current_span_type = label[2:]
            inside = True
        elif inside and label.startswith("O"):
            if current_span_type != "V":
                spans[(current_span_start,i)] = current_span_type
            inside = False
        elif inside and label.startswith("I") and label[2:] != current_span_type:            
            if current_span_type != "V":
                spans[(current_span_start,i)] = current_span_type
            inside = False
    return spans

In [259]:
def evaluate_spans(model, loader):
    """
    Span-based evaluation that properly handles padding tokens.
    """
    model.eval()
    
    total_tp = 0
    total_fp = 0
    total_fn = 0
    
    total_batches = len(loader)
    
    with torch.no_grad():
        for batch_idx, batch in enumerate(loader):
            if batch_idx % 100 == 0:
                print(f"Processing batch {batch_idx}/{total_batches}")
                
            # Process batch
            ids = batch['ids'].squeeze(1).to('cuda')
            mask = batch['mask'].squeeze(1).to('cuda')
            targets = batch['targets'].squeeze(1).to('cuda')
            pred_mask = batch['pred'].squeeze(1).to('cuda')
            
            # Get predictions
            outputs = model(ids, mask, pred_mask)
            predictions = torch.argmax(outputs, dim=2)
            
            # Process each sequence in batch
            for i in range(predictions.size(0)):
                pred_seq = predictions[i]
                target_seq = targets[i]
                attention_mask = mask[i]
                
                # Convert to labels, ignoring padding tokens
                pred_labels = []
                target_labels = []
                
                # Only process tokens where attention_mask is 1 (non-padding)
                for j in range(len(pred_seq)):
                    if attention_mask[j] == 1:  # This is a real token
                        pred_id = pred_seq[j].item()
                        target_id = target_seq[j].item()
                        
                        # Only include if target is not padding (-100)
                        if target_id != -100:
                            pred_labels.append(id_to_role[pred_id])
                            target_labels.append(id_to_role[target_id])
                
                # Extract spans for valid tokens
                pred_spans = extract_spans(pred_labels)
                target_spans = extract_spans(target_labels)
                
                # Calculate metrics
                for span, label in pred_spans.items():
                    if span in target_spans and target_spans[span] == label:
                        total_tp += 1
                    else:
                        total_fp += 1
                
                for span, label in target_spans.items():
                    if span not in pred_spans or pred_spans[span] != label:
                        total_fn += 1
            
            # Print intermediate results every 500 batches
            if batch_idx % 500 == 0 and batch_idx > 0:
                curr_p = total_tp / (total_tp + total_fp + 1e-8)
                curr_r = total_tp / (total_tp + total_fn + 1e-8)
                curr_f = 2 * curr_p * curr_r / (curr_p + curr_r + 1e-8)
                print(f"Intermediate scores at batch {batch_idx}:")
                print(f"P: {curr_p:.4f} R: {curr_r:.4f} F1: {curr_f:.4f}")
    
    # Calculate final metrics
    final_p = total_tp / (total_tp + total_fp + 1e-8)
    final_r = total_tp / (total_tp + total_fn + 1e-8)
    final_f = 2 * final_p * final_r / (final_p + final_r + 1e-8)
    
    print("\nFinal Results:")
    print(f"True Positives: {total_tp}")
    print(f"False Positives: {total_fp}")
    print(f"False Negatives: {total_fn}")
    print(f"Precision: {final_p:.4f}")
    print(f"Recall: {final_r:.4f}")
    print(f"F1: {final_f:.4f}")
    
    return final_p, final_r, final_f
print(evaluate_spans(model, loader))

Processing batch 0/43665
Processing batch 100/43665
Processing batch 200/43665
Processing batch 300/43665
Processing batch 400/43665
Processing batch 500/43665
Intermediate scores at batch 500:
P: 0.8231 R: 0.8411 F1: 0.8320
Processing batch 600/43665
Processing batch 700/43665
Processing batch 800/43665
Processing batch 900/43665
Processing batch 1000/43665
Intermediate scores at batch 1000:
P: 0.8108 R: 0.8366 F1: 0.8235
Processing batch 1100/43665
Processing batch 1200/43665
Processing batch 1300/43665
Processing batch 1400/43665
Processing batch 1500/43665
Intermediate scores at batch 1500:
P: 0.8049 R: 0.8289 F1: 0.8167
Processing batch 1600/43665
Processing batch 1700/43665
Processing batch 1800/43665
Processing batch 1900/43665
Processing batch 2000/43665
Intermediate scores at batch 2000:
P: 0.7948 R: 0.8250 F1: 0.8096
Processing batch 2100/43665
Processing batch 2200/43665
Processing batch 2300/43665
Processing batch 2400/43665
Processing batch 2500/43665
Intermediate scores a

In my evaluation, I got an F score of 0.82  (which slightly below the state-of-the art in 2018)

### OPTIONAL: 

Repeat the span-based evaluation, but print out precision/recall/f1-score for each role separately.